File: src/main/java/hiiragi283/material/compat/jei/ingredient/MaterialStackHelper.java
package hiiragi283.material.compat.jei.ingredient;

import hiiragi283.material.RMReference;
import hiiragi283.material.api.material.HiiragiMaterial;
import hiiragi283.material.api.material.MaterialStack;
import mezz.jei.api.ingredients.IIngredientHelper;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Spliterator;
import java.util.Spliterators;
import java.util.stream.StreamSupport;

public class MaterialStackHelper implements IIngredientHelper<MaterialStack> {

    @Nullable
    @Override
    public MaterialStack getMatch(@NotNull Iterable<MaterialStack> iterable, @NotNull MaterialStack materialStack) {
        return StreamSupport.stream(Spliterators.spliteratorUnknownSize(iterable.iterator(), Spliterator.ORDERED), false)
                .filter(materialStack::equalsMaterial)
                .findFirst().orElse(MaterialStack.EMPTY);
    }

    @NotNull
    @Override
    public String getDisplayName(@NotNull MaterialStack materialStack) {
        return materialStack.getTranslatedName();
    }

    @NotNull
    @Override
    public String getUniqueId(@NotNull MaterialStack materialStack) {
        return materialStack.getMaterial().map(HiiragiMaterial::name).orElse("");
    }

    @NotNull
    @Override
    public String getWildcardId(@NotNull MaterialStack materialStack) {
        return getUniqueId(materialStack);
    }

    @NotNull
    @Override
    public String getModId(@NotNull MaterialStack materialStack) {
        return RMReference.MOD_ID;
    }

    @NotNull
    @Override
    public String getResourceId(@NotNull MaterialStack materialStack) {
        return materialStack.getMaterial().map(HiiragiMaterial::name).orElse("");
    }

    @NotNull
    @Override
    public MaterialStack copyIngredient(@NotNull MaterialStack materialStack) {
        return new MaterialStack(materialStack, materialStack.amount);
    }

    @NotNull
    @Override
    public String getErrorInfo(@Nullable MaterialStack materialStack) {
        return materialStack != null ? "Errored with " + materialStack : "null";
    }

}

File: src/main/kotlin/hiiragi283/material/compat/jei/ingredients/MaterialStackHelper.kt
package hiiragi283.material.compat.jei.ingredients

import hiiragi283.material.RMReference
import hiiragi283.material.api.material.HiiragiMaterial
import mezz.jei.api.ingredients.IIngredientHelper

object MaterialStackHelper : IIngredientHelper<HiiragiMaterial> {

    override fun getMatch(iterator: MutableIterable<HiiragiMaterial>, toMatch: HiiragiMaterial): HiiragiMaterial? =
        iterator.firstOrNull { it == toMatch }

    override fun getDisplayName(material: HiiragiMaterial): String = material.getTranslatedName()

    override fun getUniqueId(material: HiiragiMaterial): String = material.toString()

    override fun getWildcardId(material: HiiragiMaterial): String = getUniqueId(material)

    override fun getModId(material: HiiragiMaterial): String = RMReference.MOD_ID

    override fun getResourceId(material: HiiragiMaterial): String = material.name

    override fun copyIngredient(material: HiiragiMaterial): HiiragiMaterial = material.copy()

    override fun getErrorInfo(material: HiiragiMaterial?): String = material?.let { "Errored with $material" } ?: "null"

}